### PR TITLE
Feat/line item support

### PR DESCRIPTION
**Pull request by fredriklagerblad**
# Jira Issue: [IPC-126] 

## What?
- Add support for line items in Underwriter, i.e. receive them from PriceEngine, store them in DB with a revision and forward to ProductPricing.

## Why?
- Part of the task to support DK taxes

## Optional checklist
- [ ] Codescouted
- [x] Unit tests written
- [x] Tested locally

[IPC-126]: https://hedvig.atlassian.net/browse/IPC-126

### CODE REVIEW

File: src/main/kotlin/com/hedvig/underwriter/service/UnderwriterImpl.kt [hunk 1]
```diff
@@ -1,38 +1,25 @@
...
 import io.micrometer.core.instrument.MeterRegistry
 import org.springframework.beans.factory.annotation.Value
 import org.springframework.stereotype.Component
 import org.springframework.stereotype.Service
-import java.lang.IllegalStateException
 import java.math.BigDecimal
 import java.time.Instant
-import java.util.UUID
-import javax.money.MonetaryAmount
+import java.util.*
```
Comment (AlexSJ1990) [inline src/main/kotlin/com/hedvig/underwriter/service/UnderwriterImpl.kt:22]: can we change back

File: src/main/kotlin/com/hedvig/underwriter/service/UnderwriterImpl.kt [hunk 2]
```diff
@@ -111,43 +98,50 @@ class UnderwriterImpl(
     }
 
     private fun complete(quote: Quote): Quote {
-        val price = getPriceRetrievedFromProductPricing(quote)
+        val priceQueryResponse = getPriceRetrievedFromProductPricing(quote)
         return quote.copy(
-            price = price.number.numberValueExact(BigDecimal::class.java),
-            currency = price.currency.currencyCode,
+            price = priceQueryResponse.price.number.numberValueExact(BigDecimal::class.java),
```
Comment (mightyost) [inline src/main/kotlin/com/hedvig/underwriter/service/UnderwriterImpl.kt:103]: Just realised this has bearing on my PR where we are going to reuse the price from a previous equal quote by member if any...

I need to copy the line items as well.

Sorry, just me thinking out loud...

And for this to work I will not reuse any previous price from recent quote unless there are any line items sat on it...

File: src/main/kotlin/com/hedvig/underwriter/model/QuoteDao.kt [hunk 3]
```diff
@@ -7,7 +7,7 @@ import org.jdbi.v3.sqlobject.statement.GetGeneratedKeys
 import org.jdbi.v3.sqlobject.statement.SqlQuery
 import org.jdbi.v3.sqlobject.statement.SqlUpdate
 import java.time.Instant
-import java.util.UUID
+import java.util.*
```
[the remark this answers is not in the document]
Comment (fredriklagerblad) [reply src/main/kotlin/com/hedvig/underwriter/model/QuoteDao.kt:10]: Hmm...it's IntelliJ that does this automatically. Will see if I can find that setting..

File: src/main/kotlin/com/hedvig/underwriter/model/QuoteRepositoryImpl.kt [hunk 4]
```diff
@@ -210,9 +218,11 @@ class QuoteRepositoryImpl(private val jdbi: Jdbi) : QuoteRepository {
             val dao = h.attach<QuoteDao>()
             val revs = dao.findRevisions(quote.id)
 
+            // Delete the line-items before the revisions, due to the FK
```
[the remark this answers is not in the document]
Comment (fredriklagerblad) [reply src/main/kotlin/com/hedvig/underwriter/model/QuoteRepositoryImpl.kt:221]: Foreign Key

File: src/main/kotlin/com/hedvig/underwriter/serviceIntegration/productPricing/dtos/mappers/OutgoingMapper.kt [hunk 5]
```diff
@@ -207,5 +215,11 @@ class OutgoingMapper {
             DanishHomeContentsType.RENT -> if (isStudent) DanishHomeContentLineOfBusiness.STUDENT_RENT else DanishHomeContentLineOfBusiness.RENT
             DanishHomeContentsType.OWN -> if (isStudent) DanishHomeContentLineOfBusiness.STUDENT_OWN else DanishHomeContentLineOfBusiness.OWN
         }
+
+        private fun toLineItems(lineItems: List<LineItem>): List<com.hedvig.productPricingObjects.dtos.LineItem> {
+            return lineItems.map { com.hedvig.productPricingObjects.dtos.LineItem(it.type, it.subType, it.amount) }
```
[the remark this answers is not in the document]
Comment (fredriklagerblad) [reply src/main/kotlin/com/hedvig/underwriter/serviceIntegration/productPricing/dtos/mappers/OutgoingMapper.kt:220]: Haven't seen that feature before! Sure, will do that.

**Review by mightyost — APPROVED** [review]
Neat!

What about support for returning the line items on:
- `QuoteController.createQuote()`
- `QuoteController.createQuoteFromAgreement()`
- `QuoteController.getQuote()`
- `QuoteController.updateQuoteInfo()`
- `QuoteController.getLatestQuoteFromMemberId()`
- `QuoteController.getAllQuotesFromMemberId()`
- `ExternalQuoteController.createQuote()`
- GraphQL endpoints?

File: src/main/kotlin/com/hedvig/underwriter/service/UnderwriterImpl.kt (hunk 2 above)
Comment (fredriklagerblad) [reply src/main/kotlin/com/hedvig/underwriter/service/UnderwriterImpl.kt:103]: Not sure I'm following, but willing to discuss if needed :-)

**Comment by fredriklagerblad** [comment]
> Neat!
> 
> What about support for returning the line items on:
> 
> * `QuoteController.createQuote()`
> * `QuoteController.createQuoteFromAgreement()`
> * `QuoteController.getQuote()`
> * `QuoteController.updateQuoteInfo()`
> * `QuoteController.getLatestQuoteFromMemberId()`
> * `QuoteController.getAllQuotesFromMemberId()`
> * `ExternalQuoteController.createQuote()`
> * GraphQL endpoints?

We have decided that for now, we only forward/expose the line-items when/where needed - i.e. to product-pricing/account-service.
Remaining we update when needed.

File: src/main/kotlin/com/hedvig/underwriter/service/UnderwriterImpl.kt (hunk 2 above)
Comment (mightyost) [reply src/main/kotlin/com/hedvig/underwriter/service/UnderwriterImpl.kt:116]: I just merged mine, you will get merge conflicts if rebasing. We can resolve it together if u want...

**Comment by mightyost** [comment]
> > Neat!
> > What about support for returning the line items on:
> > 
> > * `QuoteController.createQuote()`
> > * `QuoteController.createQuoteFromAgreement()`
> > * `QuoteController.getQuote()`
> > * `QuoteController.updateQuoteInfo()`
> > * `QuoteController.getLatestQuoteFromMemberId()`
> > * `QuoteController.getAllQuotesFromMemberId()`
> > * `ExternalQuoteController.createQuote()`
> > * GraphQL endpoints?
> 
> We have decided that for now, we only forward/expose the line-items when/where needed - i.e. to product-pricing/account-service.
> Remaining we update when needed.

I see that they are in fact returned in the `QuoteController.getQuote()`, that is good, then it is possible to test this feature using integration tests :)...

File: src/test/kotlin/com/hedvig/underwriter/web/QuoteControllerIntegrationTest.kt [hunk 6]
```diff
@@ -106,6 +119,7 @@ class QuoteControllerIntegrationTest {
 
         // THEN
         verify { memberService.finalizeOnboarding(any(), eq("s@hedvig.com")) }
+        assertLineItems(priceEngineLineItems, quoteSlot.captured.quotes.first().lineItems)
```
Comment (mightyost) [inline src/test/kotlin/com/hedvig/underwriter/web/QuoteControllerIntegrationTest.kt:122]: Maybe also call `QuoteController.getQuote(..)` and verify that the line items was correctly stored in db?

**Pull request merged**